app/gui/src/dashboard/providers/LocalStorageProvider.tsx

```diff
@@ -41,61 +40,106 @@ export default function LocalStorageProvider(props: LocalStorageProviderProps) {
 
 /** Exposes a property to get the shortcut registry. */
 export function useLocalStorage() {
-  return React.useContext(LocalStorageContext)
+  return useContext(LocalStorageContext)
 }
 
-export function useLocalStorageState<K extends LocalStorageKey>(
-  key: K,
-): readonly [
-  value: LocalStorageData[K] | undefined,
-  setValue: (newValue: React.SetStateAction<LocalStorageData[K] | undefined>) => void,
-]
-
-export function useLocalStorageState<K extends LocalStorageKey>(
-  key: K,
-  defaultValue: LocalStorageData[K],
-): readonly [
-  value: LocalStorageData[K],
-  setValue: (newValue: React.SetStateAction<LocalStorageData[K]>) => void,
-]
-
-/** Subscribe to Local Storage updates for a specific key. */
-export function useLocalStorageState<K extends LocalStorageKey>(
-  key: K,
-  defaultValue?: LocalStorageData[K],
-): readonly [
-  value: LocalStorageData[K] | undefined,
-  setValue: (newValue: LocalStorageData[K] | undefined) => void,
-] {
-  const { localStorage } = useLocalStorage()
-
-  const [value, privateSetValue] = React.useState<LocalStorageData[K] | undefined>(
-    () => localStorage.get(key) ?? defaultValue,
-  )
+/** Options for {@link defineLocalStorageKey}. */
+export interface DefineLocalStorageKeyOptions<Schema extends z.ZodSchema>
+  extends LocalStorageKeyMetadata {
+  readonly schema: (zod: typeof z) => Schema
+}
 
-  const setValue = useEventCallback(
-    (newValue: React.SetStateAction<LocalStorageData[K] | undefined>) => {
+/** Create a set of hooks for interacting with one specific local storage key. */
+export function defineLocalStorageKey<Schema extends z.ZodSchema>(
+  key: string,
+  options: DefineLocalStorageKeyOptions<Schema>,
+) {
+  /** The type of the value corresponding to this {@link LocalStorage} key. */
+  type Value = NonFunction & z.infer<Schema>
+
+  const { schema: makeSchema, ...metadata } = options
+  const schema = makeSchema(z)
+  LocalStorage.defineKey(key, metadata)
+
+  if ('error' in schema.safeParse(LocalStorage.getInstance().get(key))) {
+    LocalStorage.getInstance().delete(key)
+  }
+
+  const get = (localStorage: LocalStorage = LocalStorage.getInstance()) => {
+    // eslint-disable-next-line no-restricted-syntax, @typescript-eslint/no-unsafe-return
+    return localStorage.get(key) as Value | undefined
+  }
+
+  const set = (value: Value, localStorage: LocalStorage = LocalStorage.getInstance()) => {
+    localStorage.set(key, value)
+  }
+
+  const deleteKey = (localStorage: LocalStorage = LocalStorage.getInstance()) => {
+    localStorage.delete(key)
+  }
+
+  const useGet = () => {
+    const { localStorage } = useLocalStorage()
+
+    useEffect(() => {
+      if ('error' in schema.safeParse(localStorage.get(key))) {
```

Comment:
I think this won't notify subscribers about removing the item from localstorage.

Also I'm not a fan of doing this stuff on the hook level, I though the `defineLocalStorageKey` is just a wrapper that removes the need to write boilerplate stuff, and the whole logic is still in the LocalStorage component

Reply:
> I think this won't notify subscribers

notifying logic should still be in the `LocalStorage` class right now

> not a fan of doing this stuff on the hook level

true. but now the schema stuff no longer exists in the `LocalStorage` class, i guess not strongly opinionated on it though. is there a reason schema stuff should be in `LocalStorage`?

(previously it was there mostly because all the registry stuff was more or less global - so it wasn't an option at all for the schema to be not in the core class)

Reply:
I thought we're stil into keeping it global? And this is supposed to be a solution for import order

Reply:
Yeah, if we want developers to fall into the pit of success by forcing them to use `defineLocalStorageItem`, we need to put these api into the same folder and expose only the function, but not the `LocalStoage` class. But I still think that the LocalStorage should keep the schema, as it used to be, because we could have different wrappers around the class in the future and coping this logic might be a bad idea.

Reply:
btw not sure to have it in a separate module (so the class needs to be exported) but also hide it from auto import